### PR TITLE
feat(db): add new org member states

### PR DESCRIPTION
**Pull request by JoshFerge**
adds two flags, `restricted:member-limit` and `deactivated` to `OrganizationMember` that will be used in upcoming feature development for new OrgMember states.

### CODE REVIEW

File: src/sentry/models/organizationmember.py [hunk 1]
```diff
@@ -91,7 +91,13 @@ class OrganizationMember(Model):
     email = models.EmailField(null=True, blank=True, max_length=75)
     role = models.CharField(max_length=32, default=str(roles.get_default().id))
     flags = BitField(
-        flags=(("sso:linked", "sso:linked"), ("sso:invalid", "sso:invalid")), default=0
+        flags=(
+            ("sso:linked", "sso:linked"),
+            ("sso:invalid", "sso:invalid"),
+            ("restricted:member-limit", "restricted:member-limit"),
+            ("deactivated", "deactivated"),
```
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/models/organizationmember.py:98]: the sso states are represented by the 2 flags above this one, and the other states won't be represented as flags, but computed based on other fields. -- the notion of `OrganizationMemberState` will be an API level concept, not one in represented in the DB by an enum. (changed after trying it out that way / researching more).

[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/models/organizationmember.py:98]: @JoshFerge so what does the `deactivated` flag mean then? Will the flag `restricted:member-limit` have the same impact when it comes to auth?

[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/models/organizationmember.py:98]: deactivated will be same as the user not having any access to the org at all (not a member), and `restricted:member-limit` will allow them to log in to the org (but not see any pages).

**Review by scefali — APPROVED** [review]

**Review by wedamija — COMMENTED** [review]
Could you confirm that these bits aren't already in use via redash?

File: src/sentry/models/organizationmember.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/models/organizationmember.py:98]: @markstory `deactivated` would be used following an okta de-provision to support its user account semantics (and, i suppose, if we ever built the feature to deactivate users from the member list):
https://developer.okta.com/docs/concepts/scim/#delete-deprovision

for all intents and purposes, deactivated will basically function the same as an invited member with the token expired -- remove the user, no ability for an invite to be accepted.

It would be used separately from the other flags, and upon re-activating / joining other flags / states would be checked.

**Comment by JoshFerge** [comment]
@wedamija confirmed not in use via:

```
SELECT flags,
       count(*)
FROM sentry_organizationmember
GROUP BY flags
ORDER BY 1 DESC
```

File: src/sentry/models/organizationmember.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/models/organizationmember.py:98]: Is there any reason we wouldn't just delete the `OrganizationMember` row instead of deactivating? Having rows that are essentially soft-deleted complicates the model

[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/models/organizationmember.py:98]: @wedamija

>Is there any reason we wouldn't just delete the OrganizationMember row instead of deactivating? Having rows that are essentially soft-deleted complicates the model

TLDR; yes, to conform 100% to SCIM spec.

So in this particular case the deactivation is for conforming to the provisioning spec SCIM, or precisely how Okta and Azure want to handle when a user is "unassigned" on their end. In Okta's case, they expect that an unassignment simply switches the user (maps to our org member), to inactive, while azure sets inactive, then after 30 days, hard deletes.

I believe the reasoning behind Okta's etc. is that in enterprise use cases, there are many scenarios in which a user is unassigned and will eventually be re-assigned. (sabbatical, parental leave, etc.), in which it would be nice if the user re-joined and any config under their identity was still there. Additonally keeping deactivated users can help with audit log and paper trails (i think we handle this okay, but this is just their justification for apps that dont)

Now, all of that being said we don't necessarily _have_ to conform to this spec. Out of the several IDPs i looked at, Gitlab and Github will delete the user-org link when a request to set the user to inactive is recieved, so in theory, we could do the same. I imagine for official integrations we have to email IDP providers and say why some of our integration tests fail, but i'm guessing with a caveat they'll let us on their integration pages as an offically supported SCIM app.

But I figured, for the amount of extra work (i think is small), we might as well conform to the full spec. We may also find this feature useful in the future for other things (see https://gitlab.com/gitlab-org/gitlab/-/issues/22257 for some interesting ideas)

As for complicating the model, I believe the cases we have to handle are:

1. Making sure deactivated members aren't notified and aren't allowed to log in / access things.
2. Displaying deactivated members in the members page (filtered by default probably)
3. Handling when deactivated members need to be reactivated / are reinvited.


1 is handled by removing the user from the orgmember object, since we already know in our code not to let members without user objects attached in / get notifications.
2 is a serializer change and a UI tweak
3 needs to be handled by the rest of the scim implementation and there is planned work to do this.

Let me know what you think, It seemed the tradeoff is small but I am open to opinions.

[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/models/organizationmember.py:98]: (note that normal orgmember deletions will still be possible, and the default way for an admin to remove a member)

[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/models/organizationmember.py:98]: @JoshFerge What configuration do we save by keeping these rows alive? I guess their teams is one thing.

My main concern is that this becomes something we have to keep in mind whenever we operate on this model. It's already pretty complicated to understand, because it can represent a real org member, an invited member, a member asking for access and now with this we have deactivated members too. I guess since you're just deleting the user from the row it fits into the existing model somewhat... Do you re-link the user to the row via email address or something along those lines?

[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/models/organizationmember.py:98]: @wedamija I can't speak to deactivating members for other reasons, but for the case where we deactivate because of the member limit, we need to keep the rows alive. This is because when the org upgrades, all those deactivated members will automatically be unlocked.

[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/models/organizationmember.py:98]: @wedamija Yeah the answer is really not much, teams seem to be the only thing, so from our perspective its not super helpful (for SCIM teams will also be assigned automatically when orgmembers are created). I'm going to contact our IdPs we're integrating with and make sure that our integration will be fine w/ not supporting this behavior and proceed with deleting when we receive a request to set "inactive".

I will go ahead and close this PR for now and @scefali will open up a separate PR for the plan downgraded members only.
